x/crisis/handler.go

```diff
@@ -0,0 +1,56 @@
+package crisis
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+const ModuleName = "crisis"
+
+func NewHandler(k Keeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+
+		switch msg := msg.(type) {
+		case MsgVerifyInvariance:
+			return handleMsgVerifyInvariance(ctx, msg, k)
+		default:
+			return sdk.ErrTxDecode("invalid message parse in crisis module").Result()
+		}
+	}
+}
+
+func handleMsgVerifyInvariance(ctx sdk.Context, msg MsgVerifyInvariance, k Keeper) sdk.Result {
+
+	// get the initial gas consumption level, for refund comparison
+	initGas := ctx.GasMeter().GasConsumed()
+
+	found := false
+	for _, invarRoutes := range k.routes {
+		if invarRoutes.Route == msg.InvarianceRoute {
+			invarianceErr := invarRoutes.Invariant()
+			found = true
+		}
+	}
+	if !found {
+		return ErrUnknownInvariant(DefaultCodespace).Result()
+	}
+
+	if invarianceErr != nil {
+
+		// refund gas
+		finalGas := ctx.GasMeter().GasConsumed()
+		diffGas := finalGas - initGas
+		gasPrice := ctx.TxGasPrice()
+		refund, _ := gasPrice.MulDec(NewDec(int64(diffGas))).TruncateDecimal() // TODO verify if should round up
+		distrKeeper.DistributeFeePool(ctx, refund, msg.Sender)
+
+		// TODO replace with circuit breaker
+		panic(invarianceErr)
```

Comment:
I agree that we should panic. However, if we panic, there's no point in refunding the invariant-submitter here, since the state change will just be reverted. Not sure what we should do, maybe just refund by social consensus in the next genesis....

Reply:
Oh yeah lol, for the time being we'd likely revert from _before_ the invar-broken tx was submitted (meaning that the tx cost was never submitted to begin with) - but in the future when we don't panic but just pause certain message types this will become increasingly important